### PR TITLE
FIX: Sort indices of sparse matrices

### PR DESCRIPTION
**Pull request by david-cortes-intel**
## Description

ref https://github.com/uxlfoundation/scikit-learn-intelex/issues/1880

This PR forcibly sorts the indices of sparse matrices before passing them to any routines in oneDAL if they aren't already sorted, as some algorithms assume that they come sorted and produce incorrect results when they aren't.

It's unclear yet which algorithms require this and which ones do not. For efficiency reasons, this additional step should be skipped in algorithms that don't need it, but that would need to be done over different PRs later on.

~It's also unclear yet if this is something that should be done on the oneDAL side instead, so leaving it as a draft PR for now.~ From internal discussions, this is a reasonable approach to follow, so moving out of draft.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.
- [x] I have extended testing suite if new functionality was introduced in this PR.

**Performance**

- [x] I have provided justification why performance has changed or why changes are not expected.

### CODE REVIEW

**Comment by david-cortes-intel** [comment]
CI error is just a timeout.

File: sklearnex/utils/validation.py [hunk 1]
```diff
@@ -135,6 +136,14 @@ def validate_data(
         if check_y:
             assert_all_finite(next(arg), allow_nan=allow_nan, input_name="y")
 
+    if (
+        require_sparse_with_sorted_indices
+        and sp.issparse(X)
+        and hasattr(X, "has_sorted_indices")
+        and not X.has_sorted_indices
+    ):
+        X.sort_indices()
```
Comment (Vika-F) [inline sklearnex/utils/validation.py]: sort_indices() does the sorting inplace.
Is it Ok to alter the user-provided data during the compuattions?

Comment (david-cortes-intel) [reply sklearnex/utils/validation.py]: I guess it should be, because that's just an implementation detail of SciPy which doesn't alter the results of operations after modifying.

@icfaust Any comments here?

Comment (david-cortes-intel) [reply sklearnex/utils/validation.py]: For reference, it looks like sklearn also modifies sparse data in-place like that even when not having a `copy_X` argument:
https://github.com/scikit-learn/scikit-learn/blob/9b7a86fb6d45905eec7b7afd01d3ae32643c8180/sklearn/svm/_base.py#L359

.. so I guess it should be fine to do the same.

Comment (Vika-F) [reply sklearnex/utils/validation.py]: Yes, I understand that the matrix stays the same in some sense, but the initial order of the indices and values might be important to the user for some reason, and after X.sort_indices() the order would be altered.

File: sklearnex/utils/validation.py [hunk 2]
```diff
@@ -135,6 +136,14 @@ def validate_data(
         if check_y:
             assert_all_finite(next(arg), allow_nan=allow_nan, input_name="y")
 
+    if (
+        require_sparse_with_sorted_indices
+        and sp.issparse(X)
+        and hasattr(X, "has_sorted_indices")
```
Comment (ahuber21) [inline sklearnex/utils/validation.py]: Don't quite understand the `if` chain here. Wouldn't it break if we don't have the attribute `has_sorted_indices`, but the indices are unsorted? I wouldn't unnecessarily add the assumption of getting passed `scipy` matrices (which provide this attribute).

In other words, shouldn't we rather use `not getattr(X, "has_sorted_indices", False)`?

Comment (david-cortes-intel) [reply sklearnex/utils/validation.py]: It is only an issue with matrices in compressed format. Other formats like COO are converted to that and produce something with sorted indices as a result.

Comment (ahuber21) [reply sklearnex/utils/validation.py]: And those matrices don't have `has_sorted_indices`, but implicitly always have sorted indices? Then it all makes sense.

Comment (david-cortes-intel) [reply sklearnex/utils/validation.py]: They don't have the attribute, and are not guaranteed to have sorted indices either, but the conversion functions from SciPy to compressed formats from other formats always output something with sorted indices.

**Comment by david-cortes-intel** [comment]
Moved the sorting part inside the `to_table` function as suggested instead of having it in the validator.

**Comment by david-cortes-intel** [comment]
Failures appear to be due to timeouts. Waiting for a review from @icfaust after moving the sorting to `to_table`.

File: onedal/datatypes/numpy/data_conversion.cpp [hunk 3]
```diff
@@ -213,6 +216,11 @@ dal::table convert_to_table(py::object inp_obj, py::object queue, bool recursed)
     }
     else if (strcmp(Py_TYPE(obj)->tp_name, "csr_matrix") == 0 ||
              strcmp(Py_TYPE(obj)->tp_name, "csr_array") == 0) {
+        if (require_sparse_with_sorted_indices) {
+            if (!py::getattr(obj, "has_sorted_indices").cast<bool>()) {
+                py::reinterpret_borrow<py::object>(obj).attr("sort_indices")();
```
Comment (icfaust) [inline onedal/datatypes/numpy/data_conversion.cpp:221]: An absolute nitpick: inp_obj may have the info you want instead of needing to use a reinterpret_borrow.

File: sklearnex/svm/tests/test_svm.py [hunk 4]
```diff
@@ -91,3 +96,29 @@ def test_sklearnex_import_nusvr(dataframe, queue):
         _as_numpy(svc.dual_coef_), [[-1.0, 0.611111, 1.0, -0.611111]], rtol=1e-3
     )
     assert_allclose(_as_numpy(svc.support_), [1, 2, 3, 5])
+
+
+# https://github.com/uxlfoundation/scikit-learn-intelex/issues/1880
```
Comment (icfaust) [inline sklearnex/svm/tests/test_svm.py:101]: Thank you for this, this will help guide devs in the future a lot.

File: onedal/datatypes/numpy/data_conversion.cpp [hunk 5]
```diff
@@ -153,7 +153,10 @@ inline csr_table_t convert_to_csr_impl(PyObject *py_data,
     return res_table;
 }
 
-dal::table convert_to_table(py::object inp_obj, py::object queue, bool recursed) {
+dal::table convert_to_table(py::object inp_obj,
+                            py::object queue,
+                            bool recursed,
+                            bool require_sparse_with_sorted_indices) {
```
Comment (icfaust) [inline onedal/datatypes/numpy/data_conversion.cpp:159]: I'm debating whether require_sparse_with_sorted_indices should go before recursed, but I think that could be done in a follow-up when we run into a scenario when someone would want to set it as false, doesn't need to be done now.

File: onedal/datatypes/numpy/data_conversion.cpp (hunk 3 above)
Comment (david-cortes-intel) [reply onedal/datatypes/numpy/data_conversion.cpp:221]: The input is `PyObject`. Getting the attribute out of it without `py::reinterpret` would require using python C api methods instead of pybind11.

File: onedal/datatypes/numpy/data_conversion.cpp [hunk 6]
```diff
@@ -213,6 +216,11 @@ dal::table convert_to_table(py::object inp_obj, py::object queue, bool recursed)
     }
     else if (strcmp(Py_TYPE(obj)->tp_name, "csr_matrix") == 0 ||
              strcmp(Py_TYPE(obj)->tp_name, "csr_array") == 0) {
+        if (require_sparse_with_sorted_indices) {
+            if (!py::getattr(obj, "has_sorted_indices").cast<bool>()) {
```
[the remark this answers is not in the document]
Comment (icfaust) [reply onedal/datatypes/numpy/data_conversion.cpp:220]: Not a problem CoPilot, statement before does a guarantee on that (though I am not a fan that it checks based on the type name, probably could be something more exact).

[the remark this answers is not in the document]
Comment (david-cortes-intel) [reply onedal/datatypes/numpy/data_conversion.cpp:220]: Comment is wrong - code here is only reached if the input is a CSR matrix, which is guaranteed to have this attribute.
